src/fhirtypes/mixins.ts

```diff
@@ -53,27 +58,32 @@ export class HasId {
    * If the FHIRId does not match, log an error.
    * If the id is a valid name, sanitize it to a valid id and log a warning
    *
-   * @param sourceInfo - The FSH file and location that specified the id
+   * @param {FshStructure | FshCodeSystem | FshValueSet} fshDefinition - The entity who's id is being set
    */
-  validateId(sourceInfo: SourceInfo) {
-    let validId = idRegex.test(this.id);
+  validateId(fshDefinition: FshStructure | FshCodeSystem | FshValueSet) {
+    const idRule = _.findLast(
+      fshDefinition.rules,
+      rule => rule instanceof CaretValueRule && rule.caretPath === 'id'
+    ) as CaretValueRule;
+    const idToCheck = idRule ? (idRule.value as string) : this.id;
+    let validId = idRegex.test(idToCheck);
     if (!validId && nameRegex.test(this.id)) {
```

Comment:
If both the `Id:` and `^id` are invalid, but the `Id:` matches the name regex, then this will sanitize the `Id:` and log the warning that we did so.  BUT, the unsanitized `^id` is the one that will still be used in the formal resource -- so sanitizing the `Id:` is kind of pointless in that case.

I think we only want to do this if the `Id:` is invalid and was not overridden by a caret rule (i.e., it's the id that will be used in the resource definition).  So I think the condition should be:
```ts
if (!validId && idRule == null && nameRegex.test(this.id)) {
```

Reply:
Yes this makes even more sense... I kind of talked myself into a tizzy when looking at this function. Commit [`5323159`](https://github.com/FHIR/sushi/pull/1042/commits/532315959432afc53a72d5a962cd21544af38fef) makes it so that the Id isn't sanitized at all if a caret rule overwrites it.